cranelift-codegen/src/ir/function.rs

```diff
@@ -279,6 +279,13 @@ impl Function {
 
         Ok(())
     }
+
+    /// Returns true if the function may perform calls to other functions.
+    pub fn may_call(&self) -> bool {
+        // Conservative result: if there's at least one function signature referenced in this
+        // function, assume it will call.
+        !self.dfg.signatures.is_empty()
```

Comment:
Does it make sense walk the blocks and instructions in debug mode and assert that if `self.dfg.signatures` is empty, then there are no calls? That is, assert that this conservative property holds in practice.

Reply:
With `self.dfg.signatures` empty, legalizing/encoding any calls would panic, as clif doesnt know what the signature of the function is, which is necessary to know which value should be in which register.

Reply:
Agreed with @bjorn3; the only opcodes that return true for `is_call` are call and call_indirect, which both take in a mandatory signature reference.